beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -4672,6 +4743,13 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         };
         // Use a context without block root or proposer index so that both are checked.
         let mut ctxt = ConsensusContext::new(block.slot());
+
+        let consensus_block_value = self
+            .compute_beacon_block_reward(inner_block.to_ref(), Hash256::zero(), &mut state.clone())
```

Comment:
I had to clone `state` as it seems that `compute_beacon_block_reward` mutates it which causes tests to fail

Reply:
Hmm cloning the state is very slow, so we'll have to find a way to work around this. We could maybe modify the reward calculations to use different mutable values to keep track of their changes (instead of the state) _or_ allow the state to be mutated and then revert the changes (this is more sketchy).

Reply:
i dug in a bit deeper here and it seems that setting the participation flag when computing the beacon block reward was what was causing tests to fail. I added a bool `set_participation_flag` as an argument to `compute_beacon_block_reward`. When set to false it will prevent setting the participation flag. This removes the need to clone the state.

Reply:
Nice. That's definitely in the right direction, but it will result in incorrect rewards due to the validator participation not being updated during the calculation.

I think we could fix this by (always) cloning the previous & current validator participation from the state at the start of `compute_beacon_block_attestation_reward_altair`, and then using the cloned value as a scratchpad in place of `state.current_epoch_participation`. Then the `set_participation_flag` bool would not be necessary either.

Ideally we could also pass `BeaconState` by immutable reference to those rewards functions. That's a good way to prove that they don't make any changes.